src/miniscript/evaluate.rs

```diff
@@ -0,0 +1,606 @@
+//! # Interpreatation of Miniscript scripts
+//!
+//! Traits and implementations to support for interpretation of miniscripts.
+//!
+//!
+use bitcoin_hashes::sha256;
+use secp256k1;
+
+use Error;
+use miniscript::astelem::AstElem;
+use ToPublicKey;
+use secp256k1::{Secp256k1, Signature, VerifyOnly};
+use bitcoin_hashes::Hash;
+use bitcoin;
+
+///Primitives for Miniscript: a vector of public keys(sigs), hashlocks and timelocks
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct Primitives{
+    pub pks: Vec<bitcoin::PublicKey>,
```

Comment:
I think we should parameterize this structure over the public key type. Also it'd be cool to just return references into the descriptor rather than copying everything, like

```
pub struct Primitives<'desc, 'stack, P> {
    pub pks: Vec<&'desc P>,
    pub hashlocks: Vec<(&'stack [u8], sha256d::Hash)>,
    pub timelocks: Vec<u32>,
}
```
and you should also have doccomments (even trivial/redundant ones) on all the fields

Reply:
Also, I think `pks` should be a `HashSet` (which puts a `std::hash::Hash` requirement on `P`, which kinda sucks, but should be fine in practice), `hashlocks` should be a `HashMap<&'desc sha256d::Hash, &'a stack[u8]>`, and `timelocks` should also be a `HashSet`

Reply:
There are a few advantages of using references like this

1. Less allocation/copying
2. It gives the user a way to determine which part of the descriptor a public key came from, in case there are multiple instances of the same key (by casting to `*const _` and comparing the pointers)
3. It improves documentation because it's clear where each piece of data comes from

Reply:
Actually maybe we want to change the API entirely to

```
/// A satisfied Miniscript condition (e.g. signature requirement)
enum SatisfiedConstraint<'desc, 'stack, P> {
    PublicKey {
        key: &'desc P,
        sig: secp256k1::Signature,
    },
    HashLock {
        hash: &'desc sha256d::Hash,
        preimage: &'stack [u8],
    },
    RelativeTimeLock {
        time: &'desc u32,
    },
}
```
then
```
/// An iterator over all the constraints satisfied by a given
/// descriptor/scriptSig/witness stack tuple
pub struct SatisfiedConstraints {
    // ...
}

impl<'desc, 'stack, P> Iterator for SatisfiedConstraints {
    type Item = SatisfiedConstraint<'desc, 'stack, P>;

    fn next(&mut self) -> Option<SatisfiedConstraint<'desc, 'stack, P>> {
        // ...
    }
}
```
Does this make sense?

Reply:
Then `Descriptor::interpret` could be replaced by a function `SatisfiedConstraints::from_descriptor<'desc, 'stack>(descriptor: &'desc Descriptor, stack: Vec<StackElement<'stack>>)` which creates this iterator, (and all of the code could exist in its own module, perhaps reexported as part of the public `descriptor` module).